src/rsa/bigint.rs

```diff
@@ -39,31 +39,69 @@ impl Positive {
     // Turns a sequence of big-endian bytes into a Positive Integer.
     pub fn from_be_bytes(input: untrusted::Input)
                          -> Result<Positive, error::Unspecified> {
-        // Reject empty inputs.
-        if input.len() == 0 {
-            return Err(error::Unspecified);
-        }
         // Reject leading zeros. Also reject the value zero ([0]) because zero
         // isn't positive.
         if untrusted::Reader::new(input).peek(0) {
             return Err(error::Unspecified);
         }
+        Self::from_be_bytes_padded(input)
+    }
+
+    pub fn from_be_bytes_padded(input: untrusted::Input)
+                                -> Result<Positive, error::Unspecified> {
+        // Reject empty inputs.
+        if input.len() == 0 {
+            return Err(error::Unspecified);
+        }
         let value = unsafe {
             GFp_BN_bin2bn(input.as_slice_less_safe().as_ptr(), input.len(),
                           core::ptr::null_mut())
         };
         if value.is_null() {
             return Err(error::Unspecified);
         }
-        Ok(Positive { value: value })
+        let r = Whole(value);
+        if r.is_zero() {
+            return Err(error::Unspecified);
+        }
+        Ok(Positive(r))
     }
 
-    pub unsafe fn as_ref<'a>(&'a self) -> &'a BIGNUM { &*self.value }
+    pub fn as_ref<'a>(&'a self) -> &'a BIGNUM { self.0.as_ref() }
+
+    pub fn into_elem<F: Field>(mut self, m: &Modulus<F>)
+                               -> Result<Elem<F>, error::Unspecified> {
+        let cmp = unsafe {
+            GFp_BN_cmp(self.as_ref(), GFp_BN_MONT_CTX_get0_n(m.as_ref()))
+        };
+        if !(cmp < 0) {
+            return Err(error::Unspecified);
+        }
+        try!(bssl::map_result(unsafe {
+            GFp_BN_to_mont(self.0.as_mut_ref(), self.as_ref(), m.as_ref())
+        }));
+        Ok(Elem {
+            value: self.0,
+            field: PhantomData,
+        })
+    }
+
+    pub fn into_elem_decoded<F: Field>(self, m: &Modulus<F>)
+            -> Result<ElemDecoded<F>, error::Unspecified> {
+        let cmp = unsafe {
```

Comment:
eventually `Modulus` will be defined as something like:

```rust
struct Modulus<M> {
    n0: Limb,
    rr: Elem<M>, // Not zero or one.
    value: OddPositive,
}
```
So eventually we'll do what you're suggesting.